### ISSUE TITLE
Server not visible on network after phone screen switched off for a while

### ISSUE THREAD

**Issue by elParaguayo** [ppareit/swiftp#95]
Hi there,

I'm really hoping this app will be able to do what I need. I'm currently automating transcoding of recorded TV on my TVHeadend server and then transferring to my phone.

To do this, I need the server to be on constantly when I'm on my home wifi to allow my script to discover my phone and transfer the file. 

I scan the network for the ftp port and retrieve the IP address of the device that has this port open. This script works fine when the screen is active but once the screen is off for a certain amount of time it appears that the port is no longer open.

I'm currently using the Demo version (but will happily pay if this works) and have ticked the "Keep device awake (full CPU speed)" option.

I've also disabled power saving options for this app on my phone (Galaxy S8 Android 7.0).

Any ideas?

**Comment by ashutoshgngwr** [comment]
After ticking "Keep device awake (full CPU speed)" option, the screen of your device shouldn't turn off automatically. You'd have to manually turn it off.

And once you turn it off manually, it puts the device into the sleep mode after a 'certain amount of time'. Once the device is in sleep mode, all the open ports will be closed by the OS until the device is forced to wake up.

**Comment by elParaguayo** [comment]
Ok. That makes sense.

So there's no way to have the ports open in "sleep mode"?

**Comment by ashutoshgngwr** [comment]
You'll have to modify the code that acquire wake lock, according to your needs. In this case, i think Partial Wake Lock will do the job.

Official Developer Docs on wake lock are available [here](https://developer.android.com/reference/android/os/PowerManager.html).

**Comment by elParaguayo** [comment]
Yes, that does sound like that could work. Unfortunately I wouldn't have the first clue where to add this code!

**Comment by ashutoshgngwr** [comment]
You don't need to modify anything. I just checked out the FsService#takeWakeLock() function (called just before the FTP server thread is started) and it looks like that the app will acquire partial wake lock after unchecking "Keep device awake (full CPU speed)" option.

**Comment by elParaguayo** [comment]
I was just reading through the source and was going to ask you that very question as my reading was that there should be a partial wakelock by default unless overridden.

I'll try this later tonight.

Thanks.

**Comment by elParaguayo** [comment]
Looks like there's a bug in Android (or Samsung's modifications). The WiFi is disconnecting when the phone sleeps despite checking the option to keep it on.

Thanks for your help on this.

**Comment by Vrihub** [comment]
I'm not sure if this is related (please let me know if I should open a different ticket), but I'm experiencing a similar problem.

Sometimes while I'm transferring lots of files between my cell phone and my linux box (using midnight commander's integrated FTP client), the transfer operation gets stuck (i.e. the progress bar on my computer stops) while the cell phone screen turns off automatically to save power.

As soon as I push the "home" button on the phone to turn the screen back on, the transfer resumes automagically until it finishes OK. Please note that I don't get any error message, or broken/partial files; it's just as if the "screen off" triggers some kind of "pause" command, and the "screen on" triggers a "resume",

I have the "Keep device awake" option turned on.

**Comment by ppareit** [comment]
Android version? Could also be doze mode.

Op 11 jun. 2017 1:05 p.m. schreef "Vrihub" <notifications@github.com>:

**Comment by ashutoshgngwr** [comment]
Some device manufacturers add their proprietary battery saver software to the device so it might be possible that this software is blocking other apps from acquiring full wake lock.

**Comment by Vrihub** [comment]
> Android version? Could also be doze mode.

Android 4.2.2

> Some device manufacturers add their proprietary...

Samsung phone. So, yes it might be some proprietary stuff causing the problem

**Comment by elParaguayo** [comment]
I'm not 100% convinced this is a Samsung/Android problem.

I've tried [FTPServer](https://play.google.com/store/apps/details?id=lutey.FTPServer&hl=en_GB) and this seems to allow me to connect when my phone screen has been off for an hour or so while SwiFTP doesn't.

There's a menu option saying "Choose the WAKE_LOCK mode" which is set to "always keep awake" but I don't know what lock is being set.

**Comment by elParaguayo** [comment]
I'll reopen the issue and see if I can do some further digging to pinpoint the problem.

**Comment by elParaguayo** [comment]
The FTPServer has a text interface showing activity. When stopping the app it references releasing both WakeLock and WifiLock. I was therefore hoping that the WifiLock may have been the answer.

Unsurprisingly, I see that the WifiLock is already implemented in your code!

I do have a couple of questions:
1) Is there a particular reason why both the WakeLock and WifiLocks have setReferenceCounted(false) when the default behaviour is true? I assume it's because there's only one activity that acquires the lock so it's not anticipated that there could ever be more than one.
2) What is the default WifiLock level? I note that the lock can be [created with various levels](https://developer.android.com/reference/android/net/wifi/WifiManager.html#createWifiLock(java.lang.String)) so I wonder if escalating the level may be an option.

If I can work out how to build android apps I may give this a go but I'd guess the odds of this fixing the issue would be pretty low!

**Comment by ashutoshgngwr** [comment]
There are no levels in Wifilock as mentioned in [official docs](https://developer.android.com/reference/android/net/wifi/WifiManager.WifiLock.html).

I don't know if there is any particular reason behind setting Wifilock and Wakelock non-reference counted but it doesn't really make any difference. By default, the device shouldn't fall asleep until the Wakelock is released and same applies to Wifilock.

**Comment by elParaguayo** [comment]
Thanks. Didn't think it would make a difference but no harm in asking.

I'll fork the code and have a play. Good learning experience for me if nothing else!

**Comment by elParaguayo** [comment]
Ok. So, I tried pinging my phone from my laptop with SwiFTP running. With screen on, ping times were about 50-200ms. As soon as screen goes off, ping times start to vary wildly between 50-1200ms. After 4-5 mins, ping responses stop and I can no longer connect to SwiFTP.

I thought I'd recreate the exercise with FTPServer given that this does seem to let me connect when the screen's off. My expectation was that the ping reponses wouldn't stop. I was wrong. Ping responses stopped 4-5 minutes after screen went but I could still connect via FTP.

I tried changing the SwiFTP code where the wifilock is created:
`wifiLock = manager.createWifiLock(WifiManager.WIFI_MODE_FULL_HIGH_PERF, TAG);`
but that didn't have any effect either.

**Comment by ppareit** [comment]
I will add support to ignore the battery optimization in doze mode for SwiFTP. This will involve a runtime permission request, but should handle your use case @elParaguayo 

I'll leave the issue open till it is implemented.

**Comment by elParaguayo** [comment]
Thanks. I've got Android Studio set up at home now so I can build and test.

However, I've already disabled the battery optimisation setting in the phone so SwiFTP is listed as an "unmonitored app" so I don't know if your suggestion will make a difference but I'd be very happy to try it!

**Comment by elParaguayo** [comment]
So I tried some USB debugging tonight. Here are the lines from about 5 mins after the screen goes off:
```
06-15 23:16:16.662 8359-8359/be.ppareit.swiftp D/WifiStateChangeReceiver: We are disconnected from wifi network
06-15 23:16:16.675 8359-18404/be.ppareit.swiftp D/FsService: Server is alive
06-15 23:16:31.678 8359-18404/be.ppareit.swiftp D/FsService: Server is alive
06-15 23:16:31.681 8359-18404/be.ppareit.swiftp D/WifiStateChangeReceiver$StopServerService: Wifi connection disconnected and no longer connecting, stopping the ftp server
06-15 23:16:31.695 8359-8359/be.ppareit.swiftp V/RequestStartStopReceiver: Received: be.ppareit.swiftp.ACTION_STOP_FTPSERVER
06-15 23:16:31.702 8359-8359/be.ppareit.swiftp I/FsService: onDestroy() Stopping server
06-15 23:16:31.703 8359-15161/be.ppareit.swiftp D/FsService: Thread interrupted
06-15 23:16:31.703 8359-15161/be.ppareit.swiftp I/FsService: Terminating 1 session thread(s)
06-15 23:16:31.703 8359-15161/be.ppareit.swiftp D/SessionThread: Closing data socket
06-15 23:16:31.704 8359-15161/be.ppareit.swiftp D/FsService: Exiting cleanly, returning from run()
06-15 23:16:31.704 8359-15164/be.ppareit.swiftp D/TcpListener: Exception in TcpListener
06-15 23:16:31.708 8359-8359/be.ppareit.swiftp D/FsService: serverThread join()ed ok
06-15 23:16:31.708 8359-8359/be.ppareit.swiftp I/FsService: Closing listenSocket
06-15 23:16:31.709 8359-8359/be.ppareit.swiftp D/FsService: onDestroy: Releasing wifi lock
06-15 23:16:31.710 8359-8359/be.ppareit.swiftp D/FsService: onDestroy: Releasing wake lock
06-15 23:16:31.712 8359-8359/be.ppareit.swiftp D/FsService: FTPServerService.onDestroy() finished
06-15 23:16:31.714 8359-8359/be.ppareit.swiftp D/FsNotification: onReceive broadcast: be.ppareit.swiftp.FTPSERVER_STOPPED
06-15 23:16:31.715 8359-8359/be.ppareit.swiftp D/FsNotification: Clearing the notifications
06-15 23:16:31.726 8359-8359/be.ppareit.swiftp D/FsNotification: Cleared notification
06-15 23:16:31.733 8359-8359/be.ppareit.swiftp D/NsdService: onReceive broadcast: be.ppareit.swiftp.FTPSERVER_STOPPED
06-15 23:16:31.739 8359-8359/be.ppareit.swiftp D/NsdService: onDestroy: Entered
06-15 23:16:31.745 8359-15168/be.ppareit.swiftp D/NsdService: onServiceUnregistered: SM-G950F FTP Server
06-15 23:16:31.748 8359-8359/be.ppareit.swiftp V/FsWidgetProvider: Received broadcast: be.ppareit.swiftp.FTPSERVER_STOPPED
06-15 23:16:31.752 8359-8359/be.ppareit.swiftp D/FsWidgetProvider: UpdateService start command
06-15 23:16:31.752 8359-8359/be.ppareit.swiftp D/FsService: Server is not running (null serverThread)
```
It looks like the phone shuts off wifi (even though the app should have it locked) so it stops the server. Weird.

**Comment by constambeys** [comment]
According to google doc:
https://developer.android.com/training/monitoring-device-state/doze-standby.html

Doze restrictions. The following restrictions apply to your apps while in Doze:
Network access is suspended.
The system ignores wake locks.

**Comment by elParaguayo** [comment]
Yes, but if I had already manually added the app to the whitelist then those restrictions shouldn't apply. Or, at least, that's what I'd expect to happen.

**Comment by elParaguayo** [comment]
I do agree that the symptoms point to this being caused by doze mode.

**Comment by constambeys** [comment]
Yes the app need to be whitelisted and can use network if it holds a partial wakelock

**Comment by elParaguayo** [comment]
But that's what should be happening now. I've whitelisted the app and the partial wake lock should be acquired.

However, from the log it seems that WiFi goes off so the app releases the locks which stops the app from working.

**Comment by constambeys** [comment]
This should work based on google doc but I know that many apps are having the same problem with doze.

**Comment by elParaguayo** [comment]
The strange thing is that the FTPServer app _seems_ to work: even though I can't ping the phone, I still seem to be able to connect to FTP.

Unfortunately that app seems to crash and just doesn't feel as smart as SwiFTP so I was hoping there may be a quick fix here.

Sounds like there may be a more fundamental problem at the operating system level.

**Comment by elParaguayo** [comment]
Unfortunately, I think this is a flaw in SwiFTP.

I have set up both SwiFTP and FTPServer running simultaneously on different ports.

My phone has been off for about an hour. I can connect to FTPServer but SwiFTP won't connect (times out). This would suggest that FTPServer is holding on to the correct locks but SwiFTP seems to be releasing them.

I'll do some more testing and update this comment if I find out anything more.

**Comment by ppareit** [comment]
We should probably implement REQUEST_IGNORE_BATTERY_OPTIMIZATIONS, though this comes with and extra permission. To check if an app has this permission, you can see the settings at Settings > Battery > Battery Optimization. I would prefer not to blindly add this permission to SwiFTP, this should come with a clear warning and explication for the user, see the following note from google:

>  Note: Google Play policies prohibit apps from requesting direct exemption from Power Management features in Android 6.0+ (Doze and App Standby) unless the core function of the app is adversely affected.

Of course, if something else can be done to keep the connection open, I'm all for it!

**Comment by ppareit** [comment]
@elParaguayo Could you experiment with disabling WifiStateChangeReceiver? 

in the AndroidManifest.xml, remove lines 85 til 92:
        <receiver android:name="be.ppareit.swiftp.WifiStateChangeReceiver">
            <intent-filter>
                <action android:name="android.net.wifi.STATE_CHANGE" />
            </intent-filter>
        </receiver>

        <service android:name="be.ppareit.swiftp.WifiStateChangeReceiver$StartServerService" />
        <service android:name="be.ppareit.swiftp.WifiStateChangeReceiver$StopServerService" />

or add a return in onReceive of WifiStateChangeReceiver

**Comment by ppareit** [comment]
One situation that we would probably be able to handle is the case in which the android device exits the doze mode and the wifi connection is connecting. At the moment this stops the ftp server. We should probably wait and do nothing. I'm curious at your tests.

**Comment by elParaguayo** [comment]
I had tried commenting out the line:
```sendBroadcast(new Intent(FsService.ACTION_STOP_FTPSERVER));```
in WifiStateChangeReceiver.java.

That didn't work but I may have forgotten to re-add the app to the ignore list for battery optimisations.

I'll reset the code to the current version and try your suggested changes tonight.

**Comment by elParaguayo** [comment]
> We should probably implement REQUEST_IGNORE_BATTERY_OPTIMIZATIONS, though this comes with and extra permission. To check if an app has this permission, you can see the settings at Settings > Battery > Battery Optimization. I would prefer not to blindly add this permission to SwiFTP, this should come with a clear warning and explication for the user, see the following note from google:

Does this do anything different to a user manually adding the app to the whitelist? i.e. should I add anything extra to the code when I test or is it sufficient to manually add the app?

**Comment by elParaguayo** [comment]
So, I commented out the lines in the manifest, whitelisted the app and ran it. After 5 mins with the screen off, I can't connect! Makes no sense to me at all!

**Comment by elParaguayo** [comment]
Just a thought, adb lists my currently whitelisted apps:
```mDeviceIdleWhitelist=[1000, 1001, 1201, 5002, 10002, 10007, 10026, 10037, 10042, 10045, 10048, 10050, 10056, 10059, 10095, 10123, 10132, 10152, 10169, 10174, 10175, 10179, 10194, 10201]```
but, according to adb, the partial_wake_lock is held by:
```PARTIAL_WAKE_LOCK              'FsService' (uid=10264, pid=2689, ws=null) (uid=10264, pid=2689, ws=null, pkg=be.ppareit.swiftp, elapsedTime=1014167)```
Now, I know virtually nothing about Android programming, but if the uid is the identifier for the app then it doesn't seem to be whitelisted.

Interestingly, FTPServer, which does work, is also not in this Whitelist so I'm not convinced it's the solution.

Nevertheless, I ran:
```adb shell dumpsys deviceidle whitelist +be.ppareit.swiftp```
and the uid now appeared in the whitelist.

I restarted the app, screen off for 5 mins and... it works!

Screen's off for 20 mins and it's still connecting.

**Comment by elParaguayo** [comment]
If I get time tonight I'll try uncommenting the WifiStateChangeReceiver section in the manifest and see if it still works.

**Comment by ppareit** [comment]
Is FTPServer in the whitelist while running?

**Comment by elParaguayo** [comment]
I added the app manually from the Battery Optimisation screen on my phone. That didn't work.

I the used the adb command above to whitelist the app and that did work.

**Comment by a442509097** [comment]
adb shell dumpsys battery set ac 1   //Simulates that the phone is charging, and the system is quite running in a charging state.
adb shell dumpsys battery unplug  //Simulate unplugging the charging cable
adb shell dumpsys battery reset   //reset state 
---------------------------------------------------------------------------
Make the system think it's charging

**Comment by Xavron** [comment]
If you're seeing an issue here you should try another Android version or device. Some devices just seem to be broken (the device itself and not the app). I have one with Android 7.1.1 and that device just doesn't work properly here. Whereas an Android 8 device works great as well as others. see https://github.com/ppareit/swiftp/pull/205#issuecomment-1945940162 for further info.

I'm not seeing any actual app issue at this time that can explain it as being an app issue. There could be one hiding very well but the app works perfectly fine on other devices....

Edit. Imo this issue should be closed as _the device is the problem_.

Edit: Also, I see this on custom rom Android 15 where remote works, while local does not, when the screen is off. Official Samsung rom device works just fine. I fully do not believe this to be an app issue. This is an unfortunate Android OS / device issue.
